compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/Types.java

```diff
@@ -0,0 +1,226 @@
+
+    }
+
+    public Optional<Map<String, Symbol>> typesInModule(String org, String moduleName, String version) {
+        PackageID packageID = new PackageID(Names.fromString(org), Names.fromString(moduleName),
+                Names.fromString(version));
+
+        return getTypeDefSymbolsInModule(packageID);
+    }
+
+    public static Types getInstance(CompilerContext context) {
+        Types types = context.get(TYPES_KEY);
+        if (types == null) {
+            types = new Types(context);
+        }
+
+        return types;
+    }
+
+    private Optional<Symbol> getTypeDefByName(PackageID packageID, String typeDefName) {
+        BPackageSymbol packageSymbol = packageCache.getSymbol(packageID);
+        if (packageSymbol == null) {
+            return Optional.empty();
+        }
+
+        SymbolEnv pkgEnv = symbolTable.pkgEnvMap.get(packageSymbol);
+
+        BSymbol bSymbol;
+        Name typeDef = Names.fromString(typeDefName);
+        bSymbol = symbolResolver.lookupSymbolInMainSpace(pkgEnv, typeDef);
+
+        if (bSymbol == symbolTable.notFoundSymbol) {
+            bSymbol = findByNameInImportedPackages(packageSymbol.imports, typeDef);
+        }
+
+        if (isValidTypeDef(bSymbol)) {
+            return Optional.ofNullable(symbolFactory.getBCompiledSymbol(bSymbol, typeDefName));
+        }
+
+        return Optional.empty();
+    }
+
+    private BSymbol findByNameInImportedPackages(List<BPackageSymbol> imports, Name typeDef) {
+        for (BPackageSymbol packageSymbol : imports) {
+            SymbolEnv pkgEnv = symbolTable.pkgEnvMap.get(packageSymbol);
+            BSymbol bSymbol = symbolResolver.lookupSymbolInMainSpace(pkgEnv, typeDef);
+            if (bSymbol != symbolTable.notFoundSymbol) {
+                return bSymbol;
+            }
+        }
+
+        return symbolTable.notFoundSymbol;
+    }
+
+    private Optional<Map<String, Symbol>> getTypeDefSymbolsInModule(PackageID packageID) {
+        BPackageSymbol packageSymbol = packageCache.getSymbol(packageID);
+        if (packageSymbol == null) {
+            return Optional.empty();
+        }
+
+        SymbolEnv pkgEnv = symbolTable.pkgEnvMap.get(packageSymbol);
+        Scope pkgEnvScope = pkgEnv.scope;
+        if (pkgEnvScope != null && pkgEnvScope.entries != null) {
+            Map<String, Symbol> typeDefSymbols = new HashMap<>();
+            for (Scope.ScopeEntry scopeEntry : pkgEnvScope.entries.values()) {
+                BSymbol bSymbol = scopeEntry.symbol;
+                if (isValidTypeDef(bSymbol)) {
+                    String typeDefName = bSymbol.getName().getValue();
+                    typeDefSymbols.put(typeDefName, symbolFactory.getBCompiledSymbol(bSymbol, typeDefName));
+                }
+            }
+
+            for (BPackageSymbol importedPkgs : packageSymbol.imports) {
+                findTypeDefsInPackageSymbol(importedPkgs, typeDefSymbols);
+            }
+
+            return Optional.of(typeDefSymbols);
```

Comment:
We shouldn't be returning mutable values. Not done anywhere in the Semantic API.